From 1383c7de6e9471f7bdbb58f280760e0b26eb439b Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:36 -0500
Subject: [drm] i915: Fix lock dropping in intel_tv_detect()

Message-id: <1418852380-13061-26-git-send-email-rclark@redhat.com>
Patchwork-id: 102162
O-Subject: [RHEL7 drm 25/89] drm/i915: Fix lock dropping in intel_tv_detect()
Bugzilla: 1153301
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1153301
Upstream: since v3.16.4

commit 3bcf9780de504719929918efb2155fe47251da8d

Author:     Ville Syrjala <ville.syrjala@linux.intel.com>
AuthorDate: Tue Sep 2 12:57:22 2014 +0300
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:14:59 2014 -0500

    drm/i915: Fix lock dropping in intel_tv_detect()

    commit bbfb44e8b688e778964275ab0862f67463ba4f84 upstream.

    When intel_tv_detect() fails to do load detection it would forget to
    drop the locks and clean up the acquire context. Fix it up.

    This is a regression from:
     commit 208bf9fdcd3575aa4a5d48b3e0295f7cdaf6fc44
     Author: Ville Syrjala <ville.syrjala@linux.intel.com>
     Date:   Mon Aug 11 13:15:35 2014 +0300

        drm/i915: Fix locking for intel_enable_pipe_a()

    v2: Make the code more readable (Chris)
    v3: Drop WARN_ON(type < 0) (Chris)

    Cc: Tibor Billes <tbilles@gmx.com>
    Reported-by: Tibor Billes <tbilles@gmx.com>
    Tested-by: Tibor Billes <tbilles@gmx.com>
    Reviewed-by: Chris Wilson <chris@chris-wilson.co.uk>
    Signed-off-by: Ville Syrjala <ville.syrjala@linux.intel.com>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_tv.c b/drivers/gpu/drm/i915/intel_tv.c
index 71aa61b..de8ccab 100644
--- a/drivers/gpu/drm/i915/intel_tv.c
+++ b/drivers/gpu/drm/i915/intel_tv.c
@@ -1311,6 +1311,7 @@ intel_tv_detect(struct drm_connector *connector, bool force)
 {
  struct drm_display_mode mode;
  struct intel_tv *intel_tv = intel_attached_tv(connector);
+ enum drm_connector_status status;
  int type;
 
  DRM_DEBUG_KMS("[CONNECTOR:%d:%s] force=%d\n",
@@ -1328,16 +1329,19 @@ intel_tv_detect(struct drm_connector *connector, bool force)
   if (intel_get_load_detect_pipe(connector, &mode, &tmp, &ctx)) {
    type = intel_tv_detect_type(intel_tv, connector);
    intel_release_load_detect_pipe(connector, &tmp);
+   status = type < 0 ?
+    connector_status_disconnected :
+    connector_status_connected;
   } else
-   return connector_status_unknown;
+   status = connector_status_unknown;
 
   drm_modeset_drop_locks(&ctx);
   drm_modeset_acquire_fini(&ctx);
  } else
   return connector->status;
 
- if (type < 0)
-  return connector_status_disconnected;
+ if (status != connector_status_connected)
+  return status;
 
  intel_tv->type = type;
  intel_tv_find_better_format(connector);
-- 
1.7.1